[MAINT, packaging] Remove support for briefcase installers

jaimergp opened the pull request:
# Fixes/Closes
Closes https://github.com/napari/napari/issues/5506

# Description

The Briefcase bundles have not been working for a while (at least in some platforms), nobody is testing them either, and our development efforts are mostly directed to the `constructor` installers. 
In this PR, they are removed to [avoid CI waste](https://github.com/napari/napari/issues/5660) and permit deleting some strange code paths for Briefcase workarounds.

# References

## Type of change
- [X] Removes / deprecates code

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [x] No CI jobs for Briefcase
- [x] Tests pass without the extra Briefcase workarounds

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [x] I have made corresponding changes to the documentation: https://github.com/napari/docs/pull/147

jaimergp committed:
Remove support for briefcase installers

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jaimergp commented on napari/utils/_appdirs.py:
Many functions on this module are Briefcase-specific adjustments. We might just remove them, but I wonder if we have to follow a deprecation policy for these?

Czaki replied:
This is a private module. Did they are exposed from non-private ones?

jaimergp replied:
I didn't check so I don't know, but I honestly hope they are not being used by any other project.

Czaki replied:
Even if other project uses it but importing from a private module then it is not our problem.

jaimergp replied:
AFAIK the only user is the plugin manager dialog which we control fully, so we are good. I removed those functions because they are not needed anymore in the `PipInstaller` workarounds.

jaimergp committed:
Merge branch 'main' of github.com:napari/napari into remove-briefcase

jaimergp committed:
remove unused functions now that we don't support the briefcase workarounds

jaimergp committed:
Merge branch 'remove-briefcase' of github.com:jaimergp/napari into remove-briefcase

jaimergp committed:
deprecate running_as_bundled_app(), replace with running_as_constructor_app()

Sometimes the intent is to check whether we are in an installer-provided installation, not to check for Briefcase itself (for technical workarounds, not UX ones).

jaimergp commented on napari/_qt/dialogs/qt_package_installer.py:
```diff
@@ -130,28 +129,13 @@ def arguments(self) -> Tuple[str, ...]:
             env = QProcessEnvironment.systemEnvironment()
-        combined_paths = os.pathsep.join(
-            [
-                user_site_packages(),
```
This was only used in Briefcase, AFAIK. There's no `plugins/` installation directory anywhere else, right?

jaimergp commented on napari/_qt/menus/file_menu.py:
```diff
@@ -123,7 +123,7 @@ def __init__(self, window: 'Window') -> None:
             },
             {
-                'when': running_as_bundled_app(),
+                'when': running_as_constructor_app(),
```
Is the Restart button a UX feature or something we needed in Briefcase for XYZ reasons?

jaimergp commented on napari/_qt/qt_main_window.py:
```diff
@@ -515,7 +515,7 @@ def restart(self):
         process = QProcess()
         process.setProgram(sys.executable)
 
-        if not running_as_bundled_app():
+        if not running_as_constructor_app():
             process.setArguments(sys.argv)
```
Not sure why this is avoided in bundled apps, but 🤷 We might even _want_ the sys.argv stuff in the constructor apps. We use the `python -m napari` syntax to launch those, and those scripts are called by specific platform launchers for OS integrations... 

The broader question is... do we even need a Restart button now? It was only used with Briefcase, apparently...

jaimergp commented on napari/plugins/_plugin_manager.py:
```diff
@@ -97,9 +95,6 @@ def __init__(self) -> None:
 
-        if sys.platform.startswith('linux') and running_as_bundled_app():
-            sys.path.append(user_site_packages())
-
     def _initialize(self):
```
Bye to this hack :D `user_site_packages()` was something like `{sys.prefix}/plugins/lib/pythonX.Y/site-packages` in Briefcase, I think, but nowhere else!

jaimergp commented on napari/utils/misc.py:
```diff
@@ -51,17 +51,26 @@ def parse_version(v) -> 'packaging.version._BaseVersion':
 
 def running_as_bundled_app(*, check_conda=True) -> bool:
-    """Infer whether we are running as a briefcase bundle."""
+    """Infer whether we are running as a bundle."""
```
Now that we are deprecating this function... I guess I need to revert this?

```suggestion
    """Infer whether we are running as a briefcase bundle."""
```

Czaki commented on napari/utils/misc.py:
```diff
@@ -51,17 +51,26 @@ def parse_version(v) -> 'packaging.version._BaseVersion':
+    warnings.warn(
+        trans._(
+            "Briefcase installations are no longer supported as of vX.Y.Z. "
+            "running_as_bundled_app() will be removed in a future release.",
```
```suggestion
            "running_as_bundled_app() will be removed in a 0.6.0 release.",
```

jaimergp replied:
I'd also need advice on what to put in the `vX.Y.Z` line above :D

Czaki replied:
0.4.18 as it not working on most machines now.

jaimergp committed:
one more

jaimergp commented on napari/utils/misc.py:
```diff
@@ -51,17 +51,26 @@ def parse_version(v) -> 'packaging.version._BaseVersion':
     # From 0.4.12 we add a sentinel file next to the bundled sys.executable
+    warnings.warn(
+        trans._(
+            "Briefcase installations are no longer supported as of vX.Y.Z. "
```
```suggestion
            "Briefcase installations are no longer supported as of v0.4.18. "
```

jaimergp committed:
Apply suggestions from code review

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>
